### PR TITLE
feat: Add Evo2 fine-tuning partial-conv benchmarking

### PR DESCRIPTION
**Pull request by nvmvle**
### Description
This PR adds a benchmarking configuration for Evo2 fine-tuning with partial convolution support. The changes include:
- A new benchmarking YAML configuration file for CI/CD pipeline to test Evo2 fine-tuning performance
- Updates to the Evo2 training script to support the benchmarking workflow

### Type of changes
- [ ]  Bug fix (non-breaking change which fixes an issue)
- [x]  New feature (non-breaking change which adds functionality)
- [ ]  Refactor
- [ ]  Documentation update
- [ ]  Other (please describe):

### CI Pipeline Configuration
Configure CI behavior by applying the relevant labels:

- [SKIP_CI](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#skip_ci) - Skip all continuous integration tests
- [INCLUDE_NOTEBOOKS_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_notebooks_tests) - Execute notebook validation tests in pytest
- [INCLUDE_SLOW_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_slow_tests) - Execute tests labelled as slow in pytest for extensive testing

> [!NOTE]
> By default, the notebooks validation tests are skipped unless explicitly enabled.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

* If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
* If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Usage
```python
# Example usage of the benchmarking configuration
# The benchmarking can be triggered through CI/CD pipeline
# using the new YAML configuration at:
# ci/benchmarks/partial-conv/evo2_finetuning.yaml
```

### Pre-submit Checklist
| - [x] I have tested these changes locally
| - [ ] I have updated the documentation accordingly
| - [ ] I have added/updated tests as needed
| - [ ] All existing tests pass successfully

Signed-off-by: My Le <mvle@nvidia.com>

## Summary by CodeRabbit

* **New Features**
  * Added an optional command-line flag to clean up GPU memory before validation/inference, improving training stability on CUDA devices; the mechanism to perform this cleanup is now included.

* **Chores**
  * Added benchmark configurations to run Evo2 finetuning variants on partial-conv with preset training parameters and logging.
  * Fixed a pretraining benchmark URL token placeholder to use the correct syntax for artifact access.

### CODE REVIEW

File: ci/benchmarks/partial-conv/evo2_finetuning.yaml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      batch_size: 16
+      lr: 5e-4
+      min_lr: 5e-5
+      accumulate_grad_batches: 1
+script: |-
+  WANDB_API_KEY=$BIONEMO_WANDB_API_KEY train_${model} \
+    -d ${data_base_path}/${dataset_config} \
+    --dataset-dir=${data_base_path}/${dataset_dir} \
+    --ckpt-dir=${restore_from_checkpoint_path} \
+    --finetune \
```
Comment (jwilber) [inline ci/benchmarks/partial-conv/evo2_finetuning.yaml]: Let's remove `--finetune` here and below.

Evo 2 training and fine-tuning follow the same set of steps, so we use the same `train_evo2` command as with training Evo 2. The big difference is the `--ckpt-dir` argument which points to a pre-existing checkpoint from some other training run. Thus, to ‘finetune’ Evo 2 is just to train Evo 2 from an existing checkpoint.

That is, instead of having an explicit `--finetune` command, it's equivalent to just use `--ckpt-dir` with an existing checkpoint.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 2]
```diff
@@ -510,7 +518,27 @@ def parse_args(args: Optional[List[str]] = None) -> argparse.Namespace:
...
+            parser.error("--finetune requires --ckpt-dir to be set")
+        
+        # Always restore optimizer state when fine-tuning
+        if not args.restore_optimizer_from_ckpt:
+            args.restore_optimizer_from_ckpt = True
+            print("Fine-tuning mode: Enabling optimizer state restoration")
+        
+    # Validate checkpoint directory exists if provided
+    if args.ckpt_dir:
+        from pathlib import Path
```
Comment (jwilber) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py]: `from pathlib import Path` is already imported at the top of the file.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 3]
```diff
@@ -258,6 +258,14 @@ def parse_args(args: Optional[List[str]] = None) -> argparse.Namespace:
     )
     parser.add_argument("--seed", type=int, default=1234, help="Set random seed for training.")
     parser.add_argument("--workers", type=int, default=8, help="Number of workers to use for data loading.")
+    parser.add_argument(
```
Comment (jwilber) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py]: Actually, just remove all changes from this file `train.py`. We don't need any of them.

**Comment by jwilber** [comment]
Have we confirmed that JET has the required datasets here? I.e.
  `data_base_path`, `restore_from_checkpoint_path`, `dataset_config`, and `dataset_dir`?

If so, awesome! Then just update with my comments and it looks great

**Comment by jwilber** [comment]
Updated:

- pushed data into [mounted](https://gitlab-master.nvidia.com/clara-discovery/bionemo-github-ci/-/blob/master/jet-toolkit/jet_custom_config.yaml?ref_type=heads) locations in eos
- simplified yaml
- removed train.py changes

**Review by jwilber — APPROVED** [review]

File: sub-packages/bionemo-evo2/src/bionemo/evo2/utils/logging/callbacks.py [hunk 4]
```diff
@@ -43,6 +43,20 @@ class TEVCallback(Callback):
...
+                torch.cuda.empty_cache()
+                torch.cuda.synchronize()
+                current_device = torch.cuda.current_device()
+                torch.cuda.set_device(current_device)
+                torch.cuda.synchronize()
+                import gc
+                gc.collect()
+            except Exception as e:
+                print(f"Warning: CUDA cleanup failed: {e}")
+
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/utils/logging/callbacks.py]: I'm ok with this but can you make a new purpose built callback called something like `GarbageCollectAtInferenceTime` or something that just does this? Then add it as a new callback to the trainer?

Comment (nvmvle) [reply sub-packages/bionemo-evo2/src/bionemo/evo2/utils/logging/callbacks.py]: I created GarbageCollectAtInferenceTime, and added a param to enable/disable this callback

File: ci/benchmarks/partial-conv/evo2_finetuning.yaml [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  max_steps: 10
+  gpus: 1
+  clip_grad: 250
+  weight_decay: 0.001
+  attention_dropout: 0.01
+  hidden_dropout: 0.01
+  stop_steps: 200
+  batch_size: 2
+  variant: finetune
+  precision: fp8
```
[the remark this answers is not in the document]
Comment (nvmvle) [reply ci/benchmarks/partial-conv/evo2_finetuning.yaml:56]: `precision` needs to be presented here to pass the validation check step in pipelines

File: ci/benchmarks/partial-conv/evo2_finetuning.yaml [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  dataset_config: training_data_config.yaml
+  dataset_dir: preprocessed_data
+  val_check_interval: 5
+  seq_length: 8192
+  warmup_steps: 10
+  activation_checkpoint_layers: 2
+  lr: 0.000015
+  min_lr: 0.0000149
+  accumulate_grad_batches: 4
+  max_steps: 10
```
[the remark this answers is not in the document]
Comment (nvmvle) [reply ci/benchmarks/partial-conv/evo2_finetuning.yaml]: fixed, i increased max_steps
